### PR TITLE
Move away from ArrayComputed/ReduceComputed

### LINKED ISSUE #4831: Using Ember.computed.sort with itemController computed properties leads to and "Index out of range" error on resort
I have made a JSBin to illustrate the issue: http://emberjs.jsbin.com/zidonulu/1/edit

As you can see the default sort is working when initially displaying the output.

After clicking on the "City" button, the `gasConsumption` values are correctly updated  but the order of the list is not changed. I would expect it to change to reflect the new property values.

If you then click on "Highway" and then on "City" again an "Index out of range" error is thrown.

Is this a bug I am hitting, or am I using the sorting mechanism in a wrong way?

### LINKED ISSUE #5268: Cleanup ArrayComputed
* [ ] Don't replace items in place when property changes - #4423 
* [ ] Consider compound dependent keys - #4684

### LINKED ISSUE #5319: [BUG] arrayComputed does not recalculate after dependant array is set
Here is a failing test case: http://jsbin.com/dohavimi/1/edit?html,js,console

Basically, `computed.filterBy` or `computed.filter` fails to recompute synchronously after its dependant array is set. An explicitly defined computed property works, so it looks like an arrayComputed issue. It works if their values are not cached (if you remove the prior gets). It also works if I use `pushObjects` instead of `set`.

Am I wrong to assume the computed properties should recalculate synchronously before the final gets?

### LINKED ISSUE #5558: arrayComputed bug when sorting by fields affecting dependent computed.filterBy
http://jsfiddle.net/nx00423k/

Check/uncheck Kris Seldon.  Once you click him he should vanish, but he doesn't.  Things get really weird on subsequent clicks

This example is simple and silly, but you can imagine a more realistic example with a Ember.Select instead of a boolean field here.

---

The issue appears to be that if you:

 1. Have a Ember.computed.filter or Ember.computed.filterBy that observes and acts on a property
 2. Have a Ember.computed.sort that is looking at the same property as above

Then you can run into situations where property 2 has items that are no longer in property 1.

### LINKED ISSUE #9462: ArrayComputed issues
* [ ] Ember Data relationships and multiple computed.sort CPs can cause error - #4919
* [ ] iterDeps (ember.js#2818, canary) calls "propertyDidChange" on attributes, that were used by Ember.computed.sort - #4231
* [ ] SortableMixin does not work when whole array changes - #3706
* [ ] TypeError for Ember.computed.sort in DependentArraysObserver.prototype.dependentArrayWillChange - #5596
* [ ] Ember computed sort fails with contentIndex property - #9485

**PRs:**
* [ ] Ember.computed.sort Observers - #5150
* [ ] fix reduce computed - #9492

### LINKED ISSUE #9739: [Perf] Array property changes not being coalesced (for reduceComputed?)
See [this jsbin](http://emberjs.jsbin.com/jetevigahu/1/edit).

The `source` array changes often and in large batches. We rely on array property change coalescing for this to not be a performance issue.

The dependent `items` array has elements that each have a property `dep`, that is dependent on some property of the full `items` array - in this case, `length`. Then we have an `{{#each}}` that iterates over these dependent values.

All of this behaves fine, so far.

This breaks down, however, when you add an `Em.computed.max` on the resulting `items.dep` values. Click the button, and see how many times the `length` property gets called. 

In our codebase, the equivalent of `length` is an expensive operation that we really would have wanted and expected to run with coalesced changes, once per runloop. The expected flow is: `items` is generated, with no value for each `item.dep` yet, then `length` is computed once, then the `items.dep` values are computed, _then_ the `max` property is computed. But Ember appears to be recomputing `max` and thus forcing recomputes of everything upto `length` on every array change operation.

Thanks!

### LINKED ISSUE #10343: 'Index out of range' error when transitioning between routes
The most common error in our Ember application is this one, which presumably happens when the user transitions out of one of our routes. This happens often in production, but much rarer in development, which tells me it might be due to the number of elements in the array(?).

I thought maybe it was this bug: #9278, but it still happens with the newest version of Ember (v1.9.1).

I don't have a reliable way of triggering the error, I just have to switch back and forth a lot.

Here is a stack trace:

```
Error while processing route: product Index out of range Error: Index out of range
    at new Error (native)
    at Error.r (https://pam.no/assets/vendor-9f8fee9a7deaea9585c549a5c2cddb91.js:7:26)
    at Array.s.default.f.create.removeAt (https://pam.no/assets/vendor-9f8fee9a7deaea9585c549a5c2cddb91.js:10:31342)
    at M.removedItem (https://pam.no/assets/vendor-9f8fee9a7deaea9585c549a5c2cddb91.js:10:11031)
    at Object.m.flushChanges (https://pam.no/assets/vendor-9f8fee9a7deaea9585c549a5c2cddb91.js:10:4481)
    at Object.m.itemPropertyDidChange (https://pam.no/assets/vendor-9f8fee9a7deaea9585c549a5c2cddb91.js:10:4126)
    at Object.m.createPropertyObserver.e.observer (https://pam.no/assets/vendor-9f8fee9a7deaea9585c549a5c2cddb91.js:10:2031)
    at _ (https://pam.no/assets/vendor-9f8fee9a7deaea9585c549a5c2cddb91.js:8:7181)
    at f (https://pam.no/assets/vendor-9f8fee9a7deaea9585c549a5c2cddb91.js:7:2509)
    at v (https://pam.no/assets/vendor-9f8fee9a7deaea9585c549a5c2cddb91.js:7:27420)vendor-9f8fee9a7deaea9585c549a5c2cddb91.js:9 wvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:9 Q.errorvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:9 Cvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:14 lvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:14 i.triggervendor-9f8fee9a7deaea9585c549a5c2cddb91.js:14 (anonymous function)vendor-9f8fee9a7deaea9585c549a5c2cddb91.js:14 yvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:14 bvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:14 gvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:14 dvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:10 (anonymous function)vendor-9f8fee9a7deaea9585c549a5c2cddb91.js:4 r.invokeWithOnErrorvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:4 r.flushvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:4 n.flushvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:4 o.endvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:4 o.runvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:7 lvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:17 yt.extend.ajax.Ember.RSVP.Promise.s.successvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:2 J.Callbacks.cvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:2 J.Callbacks.h.fireWithvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:3 rvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:3 J.ajaxTransport.X.cors.e.crossDomain.send.tXMLHttpRequest.send (async)vendor-9f8fee9a7deaea9585c549a5c2cddb91.js:3 J.ajaxTransport.X.cors.e.crossDomain.sendvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:3 J.extend.ajaxvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:17 yt.extend.ajaxvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:14 _vendor-9f8fee9a7deaea9585c549a5c2cddb91.js:14 hvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:17 yt.extend.ajaxvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:17 yt.extend.findvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:16 $vendor-9f8fee9a7deaea9585c549a5c2cddb91.js:18 Ember.Object.extend.fetchRecordvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:18 rvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:18 Ember.Object.extend._flushPendingFetchForTypevendor-9f8fee9a7deaea9585c549a5c2cddb91.js:7 h.forEach.tvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:7 c.forEachvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:7 h.forEachvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:18 Ember.Object.extend.flushAllPendingFetchesvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:4 r.invokeWithOnErrorvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:4 r.flushvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:4 n.flushvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:4 o.endvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:4 o.runvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:7 lwebclient-7cb8831e571b29c1360aa0fb53e1aa73.js:4 u.extend.modelvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:8 _vendor-9f8fee9a7deaea9585c549a5c2cddb91.js:8 rvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:9 U.extend.deserializevendor-9f8fee9a7deaea9585c549a5c2cddb91.js:14 pvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:13 n.runSharedModelHookvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:14 a.getModelvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:14 rvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:14 yvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:14 bvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:14 gvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:10 (anonymous function)vendor-9f8fee9a7deaea9585c549a5c2cddb91.js:4 r.invokeWithOnErrorvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:4 r.flushvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:4 n.flushvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:4 o.endvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:4 o.runvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:4 o.joinvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:7 l.joinvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:12 d.default.y.extend._bubbleEventvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:12 (anonymous function)vendor-9f8fee9a7deaea9585c549a5c2cddb91.js:2 J.event.dispatchvendor-9f8fee9a7deaea9585c549a5c2cddb91.js:2 J.event.add.g.handlevendor-9f8fee9a7deaea9585c549a5c2cddb91.js:23 e.bugsnag
```

I'm not sure where to start looking with this one.

### LINKED ISSUE #11181: [Meta] Remove reduce computed
This PR tracks work that needs to happen to remove reduce computed 

- [ ] Implement the current RC macros in a simpler fashion where the array recomputes (there is already a PR by @wagenet ? )
- [ ] Extract the current RC into an addon

### PR DESCRIPTION
**Pull request by stefanpenner**
@wagenet  & @stefanpenner

Fixes #11181
Fixes #10343
Fixes #9739
Fixes #9462
Fixes #4919
Fixes #4231
Fixes #3706
Fixes #5596
Fixes #9485
Fixes #9492
Fixes #5319 
Fixes #5268
Fixes #4831
Fixes #5558
- [x] almost everything
- [x] computed sort
- [x] make the new macros all readOnly
- [x] double check open issues
- [x] copious peer review

### CODE REVIEW

File: packages/ember-runtime/tests/computed/reduce_computed_macros_test.js [hunk 1]
```diff
@@ -653,38 +668,43 @@ QUnit.test('it throws an error if given fewer or more than two dependent propert
...
     array1.pushObject(22);
   });
-  deepEqual(diff, [1, 2, 6, 7, 22], 'objects added to the keep array not in the remove array are added to the computed array');
+
+  deepEqual(sortedGet(obj, 'diff'), [1, 2, 6, 7, 22], "objects added to the keep array not in the remove array are added to the computed array");
 });
 
 
 function commonSortTests() {
+<<<<<<< HEAD
```
Comment (et) [inline packages/ember-runtime/tests/computed/reduce_computed_macros_test.js]: Merge conflict?

Comment (stefanpenner) [inline packages/ember-runtime/tests/computed/reduce_computed_macros_test.js]: Ya, im just going through the work to re-integrate and fix this. Figured I would get it as a PR sooner rather then later so others saw this was being worked on.

**Comment by stefanpenner** [comment]
don't forget about https://github.com/emberjs/ember.js/pull/5150

File: packages/ember-runtime/lib/computed/reduce_computed_macros.js [hunk 2]
```diff
@@ -5,18 +5,47 @@
...
+  return computed(`${dependentKey}.[]`, function() {
+    return get(this, dependentKey).reduce(callback, initialValue);
+  }).readOnly();
+}
+
+function arrayMacro(dependentKey, callback) {
+  // This is a bit ugly
+  var fullKey = dependentKey;
+  if (/@each/.test(fullKey)) {
+    dependentKey = fullKey.replace(/\.@each.*$/, '');
```
Comment (mixonic) [inline packages/ember-runtime/lib/computed/reduce_computed_macros.js]: doesn't this change the invalidation behavior of the macro in a very unexpected way?

File: packages/ember-runtime/lib/computed/reduce_computed_macros.js [hunk 3]
```diff
@@ -5,18 +5,47 @@
...
+  } else {
+    fullKey += '.[]';
+  }
+
+  return computed(fullKey, function() {
+    var value = get(this, dependentKey);
+    if (isArray(value)) {
+      return Ember.A(callback(value));
+    } else {
+      // Is this right?
```
Comment (mixonic) [inline packages/ember-runtime/lib/computed/reduce_computed_macros.js]: Should probably remove. Returning a blank array lgtm.

File: packages/ember-runtime/lib/computed/reduce_computed_macros.js [hunk 4]
```diff
@@ -5,18 +5,47 @@
...
+}
+
+function multiArrayMacro(dependentKeys, callback) {
+  var args = dependentKeys.map((key) => `${key}.[]`);
+
+  args.push(function() {
+    return Ember.A(callback.call(this, dependentKeys));
+  });
+
+  return computed.apply(this, args).readOnly().readOnly();
```
Comment (mixonic) [inline packages/ember-runtime/lib/computed/reduce_computed_macros.js]: Why two `readOnly`?

Comment (stefanpenner) [inline packages/ember-runtime/lib/computed/reduce_computed_macros.js]: wow no idea

File: packages/ember-runtime/lib/computed/reduce_computed_macros.js (hunk 2 above)
Comment (stefanpenner) [inline packages/ember-runtime/lib/computed/reduce_computed_macros.js]: unsure, this snippet is somewhat inherited. Got a test case in mind?

**Comment by mixonic** [comment]
This looks pretty great :-)

File: packages/ember-runtime/lib/computed/reduce_computed_macros.js [hunk 5]
```diff
@@ -404,41 +354,23 @@ export var filterProperty = filterBy;
...
-      if (--itemCounts[guid] === 0) {
-        array.removeObject(item);
+export function uniq(...args) {
+  return multiArrayMacro(args, function(dependentKeys) {
+    var uniq = Ember.A();
+
+    dependentKeys.forEach(dependentKey => {
+      var value = get(this, dependentKey);
+      if (isArray(value)) {
+        value.forEach(item => {
```
Comment (chadhietala) [inline packages/ember-runtime/lib/computed/reduce_computed_macros.js]: Maybe
`uniq.pushObjects(value.filter( item => uniq.indexOf(item) === -1))`

**Comment by rwjblue** [comment]
:+1: - LGTM (tested in a few apps)...

@stefanpenner - Can you prefix with `[BUGFIX beta]`?

**Comment by raytiley** [comment]
Hopefully addons are using ember-try :)

**Comment by rwjblue** [comment]
@stefanpenner - The last rebase introduced a failure.

**Comment by stefanpenner** [comment]
Thanks. Ill check it out this evening

**Comment by stefanpenner** [comment]
@rwjblue green now. This is rather large, we can briefly mention it in the meeting today to get confirmation from the rest of the team.

**Comment by kategengler** [comment]
I think this might also cover #5558

**Comment by stefanpenner** [comment]
@kategengler yup, it should totally fix that.

**Comment by dgeb** [comment]
:+1: looks good - great job, @stefanpenner!

**Comment by mmun** [comment]
A short explanation...

ArrayComputed (AC) is no longer necessary because of Glimmer's diffing based rendering. Previously, AC would meticulously track changes to arrays and keep metadata of splices (push, pop, shift, unshift, etc.) as they happened. With Glimmer's diffing, we no longer do any bookkeeping, instead all work is deferred until render and done in a single, very efficient, diffing pass. In order to do this efficiently, we use a key-based algorithm (pioneered by React).

If you don't know about keys yet, that's fine. Ember has a default key implementation (relying on Ember.guidFor) that is backwards compatible with `{{#each}}`. Semantically, diffing and ArrayComputed are identical when all keys are unique, which is the majority of cases. There is a small semantic difference when dealing with duplicate keys (which aren't supported quite yet, but soon will be), but it shouldn't affect any real world apps.

**Comment by anulman** [comment]
:+1: for better performance with fewer LOC!

**Comment by davidpett** [comment]
Thanks for the explanation @mmun, I like it and helps me understand what's going on!

**Comment by lukemelia** [comment]
Ah, how I've yearned for this day...! So good.

**Comment by cyril-sf** [comment]
#RenderEquality what a week!

**Comment by ianstarz** [comment]
The red in this diff is incredible, so great!

**Comment by toranb** [comment]
@stefanpenner great work my friend!

**Comment by taras** [comment]
move_away_from_arraycomputed_reducecomputed_by_stefanpenner_ _pull_request__11513_ _emberjs_ember_js
 sweet!

**Comment by stefanpenner** [comment]
@taras most of the additions are improved tests

**Pull request merged**